[Blazor] Adds support for preserving prerendered state in Blazor applications

javiercn opened the pull request:
The design discussion can be found [here](https://github.com/dotnet/aspnetcore/issues/28361#issuecomment-762959954)

Major points:
* The APIs are not set in stone yet, but I want to be specific and somewhat low-level to differentiate them from other more general purpose APIs that other solutions in this area might offer.
  * These APIs are for persisting serialized data of your app at different points in time.
  * We offer JSON serialization for convenience, but that's it.
* I changed the way the HTML renderer works to be shared in the scope of the request, this is somewhat orthogonal to this change but simplifies the work in prerendering a lot and we don't think there is a big behavior change.
* State preservation happens automatically when you have 1 component and requires you to use a tag helper when rendering more than one component into the page.
* I don't have tests since I'm still doing research on it, but made sure that the existing tests pass.
* I do have a sample for Blazor WebAssembly but didn't include it on the PR since it adds many files (we don't have a hosted wasm app in our samples, which I'm adding for convenience in the future).
* Some test are failing, but since I can't debug I'll resolve the issues later.

Feedback is welcome at this point.

javiercn committed:
State preservation works for Blazor server

javiercn committed:
Make state preservation work in Blazor webassembly

javiercn committed:
* Switch to byte array for the storage.
* Cleanup sources.
* Move store to SharedSource.
* Implement protected store for server-side Blazor.

javiercn committed:
Make HtmlRenderer a scoped service

javiercn committed:
Add test to validate IAsyncDisposable components get disposed during prerendering

javiercn committed:
API cleanup

javiercn committed:
Cleanups

javiercn committed:
Cleanup

javiercn committed:
Cleanups fix tests

javiercn committed:
Cleanup

javiercn committed:
Cleanup

javiercn committed:
Undo unwanted changes

javiercn commented on src/Components/Samples/BlazorServerApp/Pages/FetchData.razor:
```diff
@@ -41,6 +44,14 @@ else
+        else
+        {
+            forecasts = JsonSerializer.Deserialize<WeatherForecast[]>(data);
+        }
```
This is a sample using "the most basic" APIs to preserve the state in any way you want.

javiercn commented on src/Components/Samples/BlazorServerApp/Pages/_Host.cshtml:
```diff
@@ -2,6 +2,10 @@
 @addTagHelper *, Microsoft.AspNetCore.Mvc.TagHelpers
 
+@{
+    ViewContext.DisableAutomaticComponentPersistence(); 
```
By default the state is save automatically if you are rendering a single component on the page. You can disable this behavior on a per page basis if you are rendering multiple components or "globally" using a _viewstart.cshtml for pages in a given folder or the entire app

javiercn commented on src/Components/Web.JS/src/Services/ComponentDescriptorDiscovery.ts:
```diff
@@ -26,6 +26,33 @@ function discoverServerComponents(document: Document): ServerComponentDescriptor
+    const candidate = nodes[index];
+    const result = discoverPersistedState(candidate);
+    if(result !== ''){
+      return result;
```
We should remove the comment node from the DOM since it has served it's purpose.

javiercn committed:
Remove the state node after is found

javiercn committed:
Revert "Temporarily remove sample to make review easier"

This reverts commit 149be05e6db1e35aca57ea99b5647bf00eaf0c47.

javiercn committed:
Revert "Remove profile and sample project"

This reverts commit fd978e26cb9ba0a36ed79b90a8ed51d0ede317f4.

SteveSandersonMS commented on src/Components/Web.JS/src/Services/ComponentDescriptorDiscovery.ts:
```diff
@@ -26,6 +26,34 @@ function discoverServerComponents(document: Document): ServerComponentDescriptor
+export function discoverPersistedState(node: Node): string {
+  if (node.nodeType === Node.COMMENT_NODE) {
+    const content = node.textContent || '';
+    const stateCommentRegex = new RegExp(blazorStateCommentRegularExpression);
+    const parsedState = stateCommentRegex.exec(content);
```
I know it's a bit early to do nitpick comments, but just don't want to forget this :)

```suggestion
    const parsedState = blazorStateCommentRegularExpression.exec(content);
```

It's already a `RegExp` I think.

javiercn replied:
You are right, I think I copied this from the other comment parsing code, so they must be both wrong.

SteveSandersonMS commented on src/Components/Web.JS/src/Services/ComponentDescriptorDiscovery.ts:
```diff
@@ -26,6 +26,34 @@ function discoverServerComponents(document: Document): ServerComponentDescriptor
   return discoveredComponents.sort((a, b): number => a.sequence - b.sequence);
 }
 
+const blazorStateCommentRegularExpression = /\W*Blazor-Component-State:(?<state>[a-zA-Z0-9\+\/=]+)$/;
```
I know we have reasons to use HTML comments as a transport mechanism in other cases, but in this particular one (where we know the state doesn't have to be emitted inside an arbitrary element) I wonder if an alternative might be an actual element like `<script type="blazor-component-state">base64goo</script>` or something like that. Then we can find it directly via `document.querySelector` and don't need any regex stuff.

javiercn replied:
That sounds like a good improvement, I was simply following the established pattern that we had. Do you know if there is any issue with using an unknown script type? Would that trigger other HTML compliance tools? I'm hoping that in the proxy case it can be configured to avoid stripping something if necessary.

If you don't think there are potential browser HTML warnings then I'm happy for us to change it, it's easy to change it back later too.

javiercn replied:
updated to use the regex directly

SteveSandersonMS commented on src/Components/Web.JS/src/Boot.WebAssembly.ts:
```diff
@@ -91,6 +91,9 @@ async function boot(options?: Partial<WebAssemblyStartOptions>): Promise<void> {
   };
 
+  const persistedState = discoverPersistedState(document);
+  window['Blazor']._internal.getPersistedState = () => BINDING.js_string_to_mono_string(persistedState);
```
Maybe it's not necessary, but if we wanted, we could bypass the whole .NET string representation of the data. That is, we could do the base64-decoding on the JS side, and supply a `byte[]` to .NET via the unmarshalled interop.

This could be left as a possible future optimization.

javiercn replied:
That sounds like a good thing to do, I haven't looked at the perf just yet because it's a one time operation

javiercn replied:
Scheduled this for a follow-up PR

SteveSandersonMS commented on src/Components/Components/src/Lifetime/IComponentApplicationStateStore.cs:
```diff
@@ -0,0 +1,24 @@
+    /// <summary>
+    /// Manages the storage for components and services that are part of a Blazor application.
+    /// </summary>
+    public interface IComponentApplicationStateStore
```
This is `public`, but it would never work for a customer to implement their own, since our hosting logic relies on it being our own concrete classes.

I guess that's just the same as something like `IJSRuntime` so maybe it's fine. But if there's some way of making it not `public`, that would be easier to understand.

javiercn replied:
That is correct but in the (near) future, when we look at pausing and resuming circuits, I expect this to be the thing customers implement to define their own circuit storage mechanism.

That said, I don't think this can be made private due to the layering (MVC needs to be able to consume it)

pranavkm replied:
Do you think we could consider making it public when we're ready to do so? For now, does IVT work?

javiercn replied:
Do you have a specific concern about it? This is not something that a user will be able to use and it's not in their faces, akin to Renderer.

We don't have IVT anywhere and I would prefer if we avoid it, that said I take note to revisit this in the future as we do more work in this area.

SteveSandersonMS commented on src/Mvc/Mvc.TagHelpers/src/ComponentTagHelper.cs:
```diff
@@ -100,12 +106,44 @@ public async override Task ProcessAsync(TagHelperContext context, TagHelperOutpu
+            var htmlRenderer = requestServices.GetRequiredService<HtmlRenderer>();
             var result = await componentRenderer.RenderComponentAsync(ViewContext, ComponentType, RenderMode, _parameters);
 
+            if (ViewContext.GetAutomaticComponentPersistencePreference() && (RenderMode == RenderMode.ServerPrerendered || RenderMode == RenderMode.WebAssemblyPrerendered))
```
I'm not 100% sold on the idea of automatic persistence. There's the whole "principle of least surprise" thing and it is quite surprising that adding a second root component would - by default - deactivate the persistence of the first root component.

It would be less surprising if persistence was on if (and only if) there's a `<persist-component-state>`. That also makes it clear that you're emitting the state to the HTML. It could even be named more clearly to indicate that, e.g., `<emit-paused-component-state />`.

javiercn replied:
I was trying to be a bit smart about it and optimize for the case where there is only one component on the page. If we are ok, we can just require you to have `save-persisted-state` on the page at all times.

The goal was to make it work automagically for 1 component and force you to be explicit if you were rendering more than one component into the page.

SteveSandersonMS commented on src/Components/Components/src/Lifetime/ComponentApplicationState.cs:
```diff
@@ -0,0 +1,105 @@
+        /// Registers a callback that will be invoked when the component application is being paused.
+        /// </summary>
+        /// <param name="callback">The <see cref="Func{TResult}"/> to invoke.</param>
+        public void RegisterOnPersistingCallback(Func<Task> callback)
```
What are the use cases for this? I don't see it used in the basic example, unless I missed something.

javiercn replied:
They are in the sample I had when originally made the changes, I took the new sample project of the PR so as to not distract the review. The functionality is for services and other more complex components.

```
public class MyService
{
    public MyService(ComponentApplicationState cas)
    {
        cas.RegisterOnPersistingCallback(SaveServiceState);
    }

    private Task SaveServiceState()
    {
        ...
    }
}
```

I'll add samples in when I update the PR to address feedback and add proper tests

SteveSandersonMS commented on src/Components/Components/src/Lifetime/IComponentApplicationStateStore.cs:
```diff
@@ -0,0 +1,24 @@
+        /// Gets the persisted state from the store.
+        /// </summary>
+        /// <returns>The persisted state.</returns>
+        IDictionary<string, byte[]> GetPersistedState();
```
What's the reason for this being synchronous-only?

I know the components themselves need to be able to retrieve key-value pairs synchronously to avoid extra rendering, but I would have thought the application startup process could wait for an async loading process if it wanted.

Or is the reasoning that nobody apart from us will ever implement this interface, and we will never want to support anything that loads asynchronously? Conceptually, I could imagine some future storage that's not in the HTML file but rather goes into an external Redis store or similar.

javiercn replied:
This was an oversight on my part, it should be async

SteveSandersonMS commented on src/Shared/Components/ProtectedPrerenderComponentApplicationStore.cs:
```diff
@@ -0,0 +1,40 @@
+
+namespace Microsoft.AspNetCore.Components
+{
+    internal class ProtectedPrerenderComponentApplicationStore : PrerenderComponentApplicationStore
```
How is this feature not called [ViewState](https://docs.microsoft.com/en-us/previous-versions/aspnet/bb386448(v=vs.100))???

Joking, but almost not.

javiercn replied:
It kind of is :), but in a "good way"

SteveSandersonMS commented on src/Mvc/Mvc.TagHelpers/src/PersistComponentStateTagHelper.cs:
```diff
@@ -0,0 +1,73 @@
+        {
+            get => _persistenceMode ?? throw new InvalidOperationException("Invalid persistence mode.");
+            set => _persistenceMode = value;
+        }
```
Is there any way this could be determined automatically based on the rendermode on the `<component>` tag? Would be nice if we could avoid the risk of customers picking the wrong option and getting confused.

pranavkm replied:
Maybe we could play around with making the `component` tag a restricted child of this component and see if that helps. 

```razor
<persist-state render-mode="ServerPrerendered">
   <component type="@typeof(App)" />
</persist-state>
```

javiercn replied:
@SteveSandersonMS Yeah, we can probably keep track of the rendering modes inside the view context for this to automagically detect the mode being rendered and also throw if there are mixed server/client components since we can't determine it in that case.

javiercn replied:
@pranavkm I thought of that too, but I don't think we can do something like that since it would prevent preserving the state of multiple components rendered in different files

javiercn replied:
I pivoted this to automatically detect the render mode (with the choice to override) and removed the automatic state persistence on the single component case

SteveSandersonMS commented:
This looks like a really solid start to me! I like the feature conceptually, and the implementation and developer experience both look nice.

Looking forwards to seeing this crystallise into a finished gem.

SteveSandersonMS commented:
Just a thought - should we do anything to make it easier to deal with versioning? For example, when deploying an update to an app, customers may need to deal with some people's browsers containing state serialized from the old version and hence missing essential properties or otherwise not matching up with the new version's requirements.

I know this could be tracked within the data the developer is storing, but it's worth at least considering whether there should be a top-level mechanism to invalidate an entire store based on either a developer-supplied version number or something like a timeout (e.g., disregard any store data older than a certain timespan). Otherwise the developer has to duplicate version storage and checking in every component that uses this.

javiercn commented:
> Just a thought - should we do anything to make it easier to deal with versioning? For example, when deploying an update to an app, customers may need to deal with some people's browsers containing state serialized from the old version and hence missing essential properties or otherwise not matching up with the new version's requirements.
> 
> I know this could be tracked within the data the developer is storing, but it's worth at least considering whether there should be a top-level mechanism to invalidate an entire store based on either a developer-supplied version number or something like a timeout (e.g., disregard any store data older than a certain timespan). Otherwise the developer has to duplicate version storage and checking in every component that uses this.

That sounds like a reasonable addition, we could "TimeStamp" the state and transparently discard it if it doesn't match the running app.

I don't know if we can do something based on hashing or just something based on time (with a configurable expiration). For prerendering is not a problem, but for other implementations it might be.

I'm wondering too if this can be kept as a detail of our implementations and we don't need to expose it to customers. In the future we might want customers to make that choice based on their specific environments needs.

SteveSandersonMS commented:
> I don't know if we can do something based on hashing or just something based on time (with a configurable expiration). For prerendering is not a problem, but for other implementations it might be.

This has an interesting overlap with the hot-restart feature. When developers are editing code, if they make a rude edit, we'll re-emit the application binary and make the page reload. Do we want the preserved state to be retained in this case? Seems like a tradeoff between making a more fluid dev experience if the schemas are compatible versus the risk of weird errors if the schemas aren't compatible.

javiercn commented:
> This has an interesting overlap with the hot-restart feature. When developers are editing code, if they make a rude edit, we'll re-emit the application binary and make the page reload. Do we want the preserved state to be retained in this case? Seems like a tradeoff between making a more fluid dev experience if the schemas are compatible versus the risk of weird errors if the schemas aren't compatible.

We can talk about it, in principle there would be no preserved state since the page reload won't have a way to notify the current runtime to trigger the state preservation step.

javiercn commented:
We are in control of when the state preservation happens, so we get to decide what to implement. In principle, I don't think this feature brings up any improvement to hot reload, but do you have any scenario in mind?

SteveSandersonMS commented:
Right, yeah, for prerendering-preservation at least, there isn't any cross-page-reload state, so it's not a problem.

My gut instinct to worry about this was probably triggered by the sense that one day, we might be using this feature for more general kinds of state preservation. For example trying to handle the case where server A goes down and the user wants to reconnect to server B.

For now I agree it's not a concern.

javiercn committed:
Fix files from rebase

javiercn committed:
Address feedback

javiercn committed:
Addressed feedback

javiercn committed:
Cleanup

javiercn committed:
More tests, more cleanup

javiercn committed:
All tests work

javiercn committed:
End to end tests

javiercn commented:
🆙 📅 
### Update notes and follow ups:
* Addressed general design feedback.
* Added tests for everything.

### Major design pivots:
* Removed the idea of automatic persistence.
  * You always need to use the taghelper to update things.  
* Automatically detect the persistence mode by default with the option to override.
  * There's no need to be explicit about the component state persistence mode.

### Stuff that I'm leaving for follow-ups:
* Performance related updates:
  * Base64 decoding on JS and passing a byte array directly to wasm.
  * Putting state in a script tag instead of a comment.
* Versioning stuff:
  * I don't think we need it right now, and want to make sure we get this on people's hands.
* API refinement:
  * The current APIs are a good starting point, but I think we want more options to allow for more efficient APIs.

javiercn committed:
Update sln

javiercn committed:
Update blazor.server.js

pranavkm commented on src/Components/Components/src/Lifetime/ComponentApplicationState.cs:
```diff
@@ -0,0 +1,152 @@
+        /// <typeparam name="T">The <paramref name="instance"/> type.</typeparam>
+        /// <param name="key">The key to use to persist the state.</param>
+        /// <param name="instance">The instance to persist.</param>
+        public void PersistAsJson<T>(string key, T instance)
```
```suggestion
        public void PersistAsJson<TValue>(string key, TValue instance)
```

pranavkm commented on src/Components/Components/src/Lifetime/ComponentApplicationState.cs:
```diff
@@ -0,0 +1,152 @@
+                throw new ArgumentNullException(nameof(key));
+            }
+
+            PersistState(key, JsonSerializer.SerializeToUtf8Bytes(instance));
```
Can we use the JsonSerializerOptions here?

pranavkm commented on src/Components/Components/src/Lifetime/IComponentApplicationStateStore.cs:
```diff
@@ -0,0 +1,24 @@
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Components
```
Do these types need to appear in a top level namespace?

javiercn replied:
We can move this, the only one I want in the top level is ComponentApplicationState since that's what the components and services interact with.

I'll move this and others to the Lifetime namespace.

pranavkm commented on src/Components/Components/src/Lifetime/ComponentApplicationState.cs:
```diff
@@ -0,0 +1,152 @@
+        /// <param name="key">The key used to persist the instance.</param>
+        /// <param name="instance">The persisted instance.</param>
+        /// <returns><c>true</c> if the state was found;<c>false</c> otherwise.</returns>
+        public bool TryRetrieveFromJson<T>(string key, [MaybeNullWhen(false)] out T instance)
```
```suggestion
        public bool TryRetrieveFromJson<T>(string key, [MaybeNullWhen(false)] out T? instance)
```

pranavkm commented on src/Components/Server/src/Circuits/CircuitFactory.cs:
```diff
@@ -34,12 +36,13 @@ internal class CircuitFactory
         }
 
-        public CircuitHost CreateCircuitHost(
+        public async Task<CircuitHost> CreateCircuitHost(
```
```suggestion
        public async ValeTask<CircuitHost> CreateCircuitHostAsync(
```

pranavkm commented on src/Components/WebAssembly/Samples/HostedBlazorWebassemblyApp/Client/wwwroot/css/open-iconic/FONT-LICENSE:
```diff
@@ -0,0 +1,86 @@
+SIL OPEN FONT LICENSE Version 1.1
```
Could we have a more minimal app here? We don't really need a full-fledged app in the sandbox.

javiercn replied:
@pranavkm I chose to use the full app because otherwise in many cases is much more painful to try stuff with something that doesn't look like a template app and we only have a testasset that is close to this which is the auth project, and it's a pain to modify that one to try stuff.

SteveSandersonMS replied:
From skim reading this, it looks like the stock project template, and the app doesn't contain anything specifically to do with state preservation. So arguably it's orthogonal to this PR. Is that correct?

If so that seems OK to me. While I sympathise with @pranavkm's point, as long as this is a pure "sandbox" that isn't coupled to any particular optional features (and isn't used from E2E tests) then it seems potentially useful to have it in the solution. We *could* simplify it a bit by removing all the 3rd-party CSS stuff, but I wouldn't say that's a priority.

If it was for a test, I'd certainly argue it should be as minimal as possible to exercise the feature being tested.

javiercn replied:
Yes, I did modify the stock template a bit to make it easier to do prerendering, but that's it. It's meant for us to be able to try stuff quickly. Otherwise it takes me 5-10 minutes every time I want to try something in this space. We also have a BlazorServerApp project in the samples for that same reason.

javiercn replied:
We can always decide to remove it in the future if we don't want it any longer

pranavkm commented on src/Mvc/Mvc.ViewFeatures/src/RazorComponents/ComponentRenderer.cs:
```diff
@@ -147,4 +191,22 @@ private IHtmlContent NonPrerenderedWebAssemblyComponent(HttpContext context, Typ
         }
     }
+
+    internal class InvokedRenderModes
```
Is there a reason you need this indirection? (Could it just be the enum)?

javiercn replied:
I wanted to avoid the continuous boxing/unboxing in/out of the Items collection. Am I missing something here?

pranavkm commented on src/Mvc/Mvc.TagHelpers/src/PersistComponentStateTagHelper.cs:
```diff
@@ -0,0 +1,89 @@
+                    InvokedRenderModes.Mode.Server =>
+                        new ProtectedPrerenderComponentApplicationStore(services.GetRequiredService<IDataProtectionProvider>()),
+                    InvokedRenderModes.Mode.ServerAndWebAssembly =>
+                        throw new InvalidOperationException("The persistence mode could not be determine based on the rendered components. Specify the PersistenceMode explicitly."),
```
```suggestion
                        throw new InvalidOperationException($"The persistence mode could not be determined based on the rendered components. Specify a value using '{PersistenceModeName}' attribute."),
```

pranavkm commented on src/Mvc/Mvc.TagHelpers/src/PersistComponentStateTagHelper.cs:
```diff
@@ -0,0 +1,89 @@
+                null => ComponentRenderer.GetPersistStateRenderMode(ViewContext) switch
+                {
+                    InvokedRenderModes.Mode.None =>
+                        null,
```
What does this scenario mean?

javiercn replied:
No `<component` tag helper was used on the page, so we don't know what mode to infer and hence we avoid rendering anything on the page.

javiercn replied:
An example is if you have this th and your component is rendered within an `if`, this makes sure that we don't break and instead we simply avoid rendering anything.

pranavkm commented on src/Mvc/Mvc.TagHelpers/src/PersistComponentStateTagHelper.cs:
```diff
@@ -0,0 +1,89 @@
+            if (store != null)
+            {
+                await manager.PersistStateAsync(store, renderer);
+                output.Content.SetHtmlContent(new PersistedStateContent(store.PersistedState));
```
You could use a `HtmlContentBuilder` instead of creating a new type.

javiercn replied:
I wasn't aware of HtmlContentBuilder, I followed the pattern we have with the serializers.

javiercn replied:
I've changed this one and I'll revisit others in the future

javiercn commented on src/Components/Components/src/Lifetime/ComponentApplicationLifetime.cs:
```diff
@@ -0,0 +1,79 @@
+    /// <summary>
+    /// Manages the lifetime of a component application.
+    /// </summary>
+    public class ComponentApplicationLifetime
```
I think so. 

This is completely "backend" agnostic and can be used for "tombstoning" in general, there's nothing web specific about it.

javiercn commented on src/Components/test/E2ETest/Tests/WebAssemblyAuthenticationTests.cs:
```diff
@@ -60,6 +60,7 @@ static WebAssemblyAuthenticationTests()
         protected override void InitializeAsyncCore()
         {
             Navigate("/", noReload: true);
+            Browser.Manage().Window.Size = new Size(1024, 800);
```
This is "orthogonal" but it was causing a bug in the Auth tests, there's a separate issue tracking the fix for Identity UI

javiercn committed:
Addressed feedback and cleanup

SteveSandersonMS commented on src/Components/Components/src/Lifetime/ComponentApplicationState.cs:
```diff
@@ -0,0 +1,152 @@
+namespace Microsoft.AspNetCore.Components
+{
+    /// <summary>
+    /// The state for the component and services of a components application.
```
```suggestion
    /// The state for the components and services of a components application.
```

SteveSandersonMS commented on src/Components/Components/src/Lifetime/ComponentApplicationState.cs:
```diff
@@ -0,0 +1,152 @@
+        /// </summary>
+        /// <param name="key">The key used to persist the state.</param>
+        /// <param name="value">The persisted state.</param>
+        /// <returns><c>true</c> if the state was found;<c>false</c> otherwise.</returns>
```
```suggestion
        /// <returns><c>true</c> if the state was found; <c>false</c> otherwise.</returns>
```

javiercn committed:
Fix build issues

SteveSandersonMS commented on src/Components/Components/src/Lifetime/ComponentApplicationState.cs:
```diff
@@ -0,0 +1,152 @@
+
+            if (_existingState.TryGetValue(key, out value))
+            {
+                _existingState.Remove(key);
```
I'm concerned "auto-remove state on retrieval" is a very strange pattern that will confuse people and make things harder for them. At the very least if we retain this behavior we should change the API name to something like `TryRetrieveAndRemovePersistedState`.

The sort of case where I think this risks making things difficult is if there are multiple instances of a component that make use of the same state. This auto-removal behavior means the developer is forced to invent a system for giving each of them unique keys that they can regenerate during rehydration.

Is there a strict reason why we have to do this? What goes wrong if we leave the state in place?

javiercn replied:
I thought about this and the answer that I have is that multiple instances of a component should not use the same state. The state is private, its for you and you should not "share" it with anyone through the cache.

If you want to share a piece of data like some values you retrieved from the database, the logical option is that a service or top level component participates in the restore process and is the one responsible for restoring the data and sharing it with the rest of the components.

There a couple of drawbacks around not doing this:
* Memory is never released since it's hold forever in the cache, which is not a feature we want to have.
* The behavior is not correct when you do something like a client navigation away and back from the page, since the state is still present there.

SteveSandersonMS replied:
OK, once I saw the samples, it was a bit clearer to me why the pattern is useful:

```cs
if (!TryRetrieveAndRemoveState(key, out value))
{
    value = await GetStateFromDb();
}
```

I do think it's important that we have the word `Remove` in the API name to make clear this is the expected pattern, otherwise people are not going to find this clear. It would be like if `Dictionary` auto-removed things when you called `Get`.

As for the drawback I was concerned about with sharing states across components, I do still think some people will run into issues with that, but given you don't want the same instance of a component to reuse the stored state more than once (e.g., after client-side navigation) the only alternative I can think of is having some notion of detecting quiescence at which point we discard the state store and only use freshly-loaded data from then on. I know you had a concern about how meaningful it would be to detect quiescence, but we do it for prerendering. Is there a definite reason you think we shouldn't use the same rule to detect it after entering interactive rendering?

SteveSandersonMS replied:
I've just realised a weird hacky thing that developers are going to do to work around the restriction we're imposing here:

```
if (componentState.TryRetrieveAndRemoveState(key, out value))
{
    // Put the state back so other components can read it too
    componentState.PersistAsJson(key, value);
}
else
{
    value = await GetStateFromDb();
}
```

I don't mean to say this PR should be blocked on it, but we should decide if this is something we want to do anything about. We don't want to be imposing restrictions that have easy workarounds because then not only are our goals not being met, we're creating battles with our customers.

It's also a bit unclear, given you can do this, when exactly you are supposed to persist state. Should you do it after each time state is retrieved from a DB, just in case the application dies unexpectedly? I know that's not our goal, but developers might think that's how it works.

We could either have the `ComponentApplicationState` service have different modes it goes through (e.g., it only switches into "accepting new state" mode once the system begins persisting state), or we could pass two different services into the component: `IComponentStateProvider` and `IComponentStatePersister` which have completely separate storage - one is read-only, the other is write-only.

javiercn replied:
They can't do that, the restored state is kept separate from the current state, so even if you persist it, other components won't be able to see it.

javiercn replied:
I found a cheap way of also "Clearing" the state after the first render, so I've done that since it's better for SSB and also on Wasm for consistency. That said, I'm keeping the current behavior because I don't want to promote "sharing" of state across components.

The current approach is more conservative and enables us to relax this later on if we want without it being a breaking change.

SteveSandersonMS commented on src/Components/Components/src/Lifetime/ComponentApplicationState.cs:
```diff
@@ -0,0 +1,152 @@
+
+            if (_currentState.ContainsKey(key))
+            {
+                throw new ArgumentException($"There is already a persisted object under the same key '{key}'");
```
This also seems pretty aggressive. What if there might be multiple instances of a state-preserving component? The developer has to invent a system to give them unique keys.

Possibilities:

 * Change this to `TryPreserveState`, using the same semantics as `Dictionary.TryAdd`
 * Or at least offer a `bool HasState(key)` method so the developer can avoid causing an exception, though that's not as convenient

javiercn replied:
State should be private to each component/service, you need to account for that when you are designing the state preservation of your component, if you need to share, you can put the data in a common service.

SteveSandersonMS replied:
OK, makes sense. Thanks.

SteveSandersonMS commented on src/Components/Components/src/PublicAPI.Unshipped.txt:
```diff
@@ -1,6 +1,20 @@
+Microsoft.AspNetCore.Components.ComponentApplicationState
+Microsoft.AspNetCore.Components.ComponentApplicationState.PersistAsJson<T>(string! key, T instance) -> void
+Microsoft.AspNetCore.Components.ComponentApplicationState.PersistState(string! key, byte[]! value) -> void
+Microsoft.AspNetCore.Components.ComponentApplicationState.RegisterOnPersistingCallback(System.Func<System.Threading.Tasks.Task!>! callback) -> void
```
Should there be a corresponding `RemoveOnPersistingCallback`? Otherwise it seems like people would be forced to leak memory if they are doing this from some component or service that has a lifetime shorter than the application itself.

javiercn replied:
That's a good suggestion, I'll add it

javiercn replied:
I've switched this to be a full fledged event. I didn't like the name `OnPersisting` (which is what I've chosen so far) so I'm extremely open to alternative names. I want to avoid generic terms like PersistingApplicationState and similar, since I find them too generic and likely overloaded

SteveSandersonMS commented on src/Components/Components/test/Lifetime/ComponentApplicationLifetimeTest.cs:
```diff
@@ -0,0 +1,158 @@
+        }
+
+        [Fact]
+        public async Task PersistStateAsync_InvokesPauseCallbacksDuringPersist()
```
I'm not sure about this, but is it worth adding a test describing something about the error handling behavior? If one of the callbacks throws, do the subsequent ones run? Are there are any guarantees about ordering?

Maybe you don't want to define the order, which is fine. And maybe the error handling rule is "if one of them throws, all subsequent behavior is undefined", which is fine *if* we have no goals to make this scenario recoverable in any way (e.g., persisting as much state as we can).

javiercn replied:
This is a good point. I think it can be worth treating this as "best effort". State is private to each component and a component/service must be able to function without their state (state is just a perf optimization)

SteveSandersonMS commented on src/Components/Web.JS/src/Boot.WebAssembly.ts:
```diff
@@ -91,6 +91,9 @@ async function boot(options?: Partial<WebAssemblyStartOptions>): Promise<void> {
     getParameterValues: (id) => BINDING.js_string_to_mono_string(componentAttacher.getParameterValues(id) || ''),
   };
 
+  const persistedState = discoverPersistedState(document);
+  window['Blazor']._internal.getPersistedState = () => BINDING.js_string_to_mono_string(persistedState);
```
Could we defer the scan through the document until the .NET side asks for it? Otherwise everyone has to pay the cost even if they aren't using the feature. I know this is not a massively expensive process but sooner or later we're going to have to streamline the startup process as much as we can, so would be good to avoid adding more things to it if it's simple to avoid them.

```suggestion
  window['Blazor']._internal.getPersistedState = () => BINDING.js_string_to_mono_string(discoverPersistedState(document));
```

SteveSandersonMS commented on src/Components/Web.JS/src/Boot.Server.ts:
```diff
@@ -34,7 +34,9 @@ async function boot(userOptions?: Partial<CircuitStartOptions>): Promise<void> {
-  const circuit = new CircuitDescriptor(components);
+  const appState = discoverPersistedState(document);
+  const circuit = new CircuitDescriptor(components, appState);
+
```
```suggestion
```

SteveSandersonMS commented on src/Components/Web.JS/src/Services/ComponentDescriptorDiscovery.ts:
```diff
@@ -26,6 +26,35 @@ function discoverServerComponents(document: Document): ServerComponentDescriptor
   return discoveredComponents.sort((a, b): number => a.sequence - b.sequence);
 }
 
+const blazorStateCommentRegularExpression = /\W*Blazor-Component-State:(?<state>[a-zA-Z0-9\+\/=]+)$/;
```
```suggestion
const blazorStateCommentRegularExpression = /^\s*Blazor-Component-State:(?<state>[a-zA-Z0-9\+\/=]+)$/;
```

The `^` at the start will let it detect non-matches massively faster. For the `\W`, was this meant to match "whitespace" or all non-word characters? Any reason why we'd need to allow other non-word characters?

SteveSandersonMS commented on src/Components/Web.JS/src/Services/ComponentDescriptorDiscovery.ts:
```diff
@@ -26,6 +26,35 @@ function discoverServerComponents(document: Document): ServerComponentDescriptor
 
+const blazorStateCommentRegularExpression = /\W*Blazor-Component-State:(?<state>[a-zA-Z0-9\+\/=]+)$/;
+
+export function discoverPersistedState(node: Node): string {
```
Could this return `string | null`? I'm not sure why we're coalescing with `''` below - it's odd to use empty string as the absence of a value.

javiercn replied:
I've changed it, the reason was that it got passed to mono_js_string_to_string and that requires a string and it was less hassle to just return a value and avoid null/undefined all-together

SteveSandersonMS commented on src/Components/Web.JS/src/Services/ComponentDescriptorDiscovery.ts:
```diff
@@ -26,6 +26,35 @@ function discoverServerComponents(document: Document): ServerComponentDescriptor
+  for (let index = 0; index < nodes.length; index++) {
+    const candidate = nodes[index];
+    const result = discoverPersistedState(candidate);
+    if(result && result !== ''){
```
An empty string is falsey, so the `result !== ''` part is redundant. In any case if we change this to use `null` instead of empty string for "no match found" this can still be just `if (result)`.

SteveSandersonMS commented on src/Components/WebAssembly/Samples/HostedBlazorWebassemblyApp/Server/Startup.cs:
```diff
@@ -0,0 +1,62 @@
+            {
+                endpoints.MapRazorPages();
+                endpoints.MapControllers();
+                //endpoints.MapFallbackToFile("index.html");
```
```suggestion
```

javiercn replied:
I'm leaving that there so that it's easy to switch back to non-prerendering when we want to test something quickly.

SteveSandersonMS commented on src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHost.cs:
```diff
@@ -131,6 +141,13 @@ internal async Task RunAsyncCore(CancellationToken cancellationToken)
             // This is the earliest opportunity to fetch satellite assemblies for this selection.
             await CultureProvider.LoadCurrentCultureResourcesAsync();
 
+            var manager = Services.GetRequiredService<ComponentApplicationLifetime>();
```
How would it look if we wanted to make this feature linkable-out? Instead of baking it directly into `WebAssemblyHost`, could we make the initialization something the developer calls themselves from their `Program.cs` logic? So if they didn't have that call, there wouldn't be any references to all these types and the whole thing would disappear.

I know we could argue that this specific feature isn't going to single-handedly make a giant difference, but we do need to be in the habit of designing new features for linkability so that (1) it doesn't grow too fast as we add features, and (2) we don't have to work so hard to reason about how each extra system API we reference while maintaining existing code will balloon the output for everyone.

I know this would be a disruptive change right now. We have a bit more time before the preview 2 deadline so it probably is possible to look into in the short term, but I'm OK with tracking it as a follow-up if you prefer.

javiercn replied:
I'm glad that you bring this up.

Initializations like these have been building up across our hosts (prerender, wasm, and server). I have a design in mind to "externalize" them and make them a general concept that third-party libraries can also leverage. 

I'm not going to deal with this in this PR, but I'll chat with you about it in the following weeks.

javiercn replied:
I filed an issue so that we don't loose track of this, but this is what I'm thinking about https://github.com/dotnet/aspnetcore/issues/30304

SteveSandersonMS commented on src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostBuilder.cs:
```diff
@@ -229,6 +237,8 @@ internal void InitializeDefaultServices()
             Services.AddSingleton<INavigationInterception>(WebAssemblyNavigationInterception.Instance);
             Services.AddSingleton(new LazyAssemblyLoader(DefaultWebAssemblyJSRuntime.Instance));
+            Services.AddSingleton<ComponentApplicationLifetime>();
+            Services.AddSingleton<ComponentApplicationState>(sp => sp.GetRequiredService<ComponentApplicationLifetime>().State);
```
As per comment above, would be great to see this replace with a call like `host.Services.AddComponentStatePreservation()` in `Program.cs`.

javiercn replied:
As per my comment [here](https://github.com/dotnet/aspnetcore/pull/29717#discussion_r579280920) are you ok if we leave it as-is for now and we revisit? The initialization logic is a bit "intricate" and don't want to add it to this PR

SteveSandersonMS commented on src/Mvc/Mvc.TagHelpers/src/PersistenceMode.cs:
```diff
@@ -0,0 +1,21 @@
+        Server,
+
+        /// <summary>
+        /// The state is persisted for a Blazor webassembly application.
```
```suggestion
        /// The state is persisted for a Blazor WebAssembly application.
```

SteveSandersonMS commented on src/Mvc/Mvc.TagHelpers/src/PersistenceMode.cs:
```diff
@@ -0,0 +1,21 @@
+    public enum PersistenceMode
+    {
+        /// <summary>
+        /// The state is persisted for a Blazor server application.
```
```suggestion
        /// The state is persisted for a Blazor Server application.
```

SteveSandersonMS commented on src/Mvc/Mvc.TagHelpers/src/Resources.resx:
```diff
@@ -165,4 +165,7 @@
     <value>A value for the '{0}' attribute must be supplied to the '{1}' tag helper.</value>
   </data>
+  <data name="PersistComponentStateTagHelper_FailedToInferComponentPersistenceMode" xml:space="preserve">
+    <value>The persistence mode could not be determined based on the rendered components. Specify a value using '{0}' attribute.</value>
```
```suggestion
    <value>The persistence mode could not be determined based on the rendered components. Specify a value using the '{0}' attribute.</value>
```

javiercn committed:
Cleanups and feedback

javiercn committed:
Follow-up feedback

javiercn committed:
Fix build break

javiercn commented:
🆙 📅 

Most important changes:
* Run all pause callbacks at the same time instead of sequentially.
* Handle errors in callbacks.
* Clear state after the initial render for our components.
* Renamed stuff in ComponentApplicationState.
  * Major change is TryRetrievePersistedState -> TryRedeemPersistedState. I've also added clarifications on the docs to make clear that it removes the item when you retrieve it.

Addressed feedback comments

javiercn committed:
Fix broken build

javiercn committed:
Fix unit test
